Validation/HGCalValidation/test/HGCalWaferHitCheck.cc

```diff
@@ -0,0 +1,263 @@
+// system include files
+#include <cmath>
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "DataFormats/ForwardDetId/interface/HFNoseDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCScintillatorDetId.h"
+#include "DataFormats/HGCDigi/interface/HGCDigiCollections.h"
+#include "DataFormats/HGCRecHit/interface/HGCRecHitCollections.h"
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDAnalyzer.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/HGCalCommonData/interface/HGCalDDDConstants.h"
+#include "Geometry/HGCalCommonData/interface/HGCalWaferIndex.h"
+
+#include "SimDataFormats/CaloHit/interface/PCaloHitContainer.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+
+namespace HGCalValidSimhitCheck {
+  struct Counters {
+    Counters() {
+      badTypes_.clear();
+      occupancy_.clear();
+      goodChannels_.clear();
+    }
+    mutable std::map<int, int> badTypes_, occupancy_;
+    mutable std::vector<int> goodChannels_;
+  };
+
+  static std::mutex mtx_;
```

Comment:
This is too eager: now two separate `HGCalWaferHitCheck` modules will share the same mutex, while it would be sufficient for them to have their own. I think it would be clearest to move the mutex as part of `Counters`, because that's where the data to be protected is as well.

Reply:
When I put mtx_ a part of Counters and I use in endStream globalCache_->mtx_.lock() or unlock() I get this message: "undefined reference to HGCalValidSimhitCheck::Counters::mtx_" in ld

Reply:
Please suggest!

Reply:
Only if I declare mtx_ mutable instead static it works

Reply:
Please show code and full error.

Reply:
Is it OK to have it that way?

Reply:
Committed the code - this one links fine

Reply:
But if I change mtx_ from mutable to static there is a link error of undefined reference

Reply:
`mutable` and non-`static` is exactly what it should be.